crates/volta-core/src/run/mod.rs

```diff
@@ -0,0 +1,130 @@
+//! Types and helpers for executing command-line tools.
+
+use std::env::{self, args_os, ArgsOs};
+use std::ffi::{OsStr, OsString};
+use std::path::Path;
+use std::process::{Command, ExitStatus};
+
+use crate::command::create_command;
+use crate::env::UNSAFE_GLOBAL;
+use crate::error::ErrorDetails;
+use crate::path;
+use crate::platform::System;
+use crate::session::Session;
+use crate::signal::pass_control_to_shim;
+use volta_fail::{Fallible, ResultExt};
+
+mod binary;
+mod node;
+mod npm;
+mod npx;
+mod yarn;
+
+pub use binary::DefaultBinary;
+
+/// Distinguish global `add` commands in npm or yarn from all others.
+enum CommandArg {
+    /// The command is a *global* add command.
+    GlobalAdd(Option<OsString>),
+    /// The command is a local, i.e. non-global, add command.
+    NotGlobalAdd,
+}
+
+pub fn execute_tool(session: &mut Session) -> Fallible<ExitStatus> {
+    path::ensure_volta_dirs_exist()?;
+
+    let mut args = args_os();
+    let exe = get_tool_name(&mut args)?;
+
+    let command = match &exe.to_str() {
+        Some("node") => node::command(args, session)?,
+        Some("npm") => npm::command(args, session)?,
+        Some("npx") => npx::command(args, session)?,
+        Some("yarn") => yarn::command(args, session)?,
+        _ => binary::command(exe, args, session)?,
+    };
+
+    command.exec()
+}
+
+/// Represents the command to execute a tool
+struct ToolCommand {
+    command: Command,
+    error: ErrorDetails,
```

Comment:
Yeah, that's essentially what it does. It's mostly for the pass-through behavior, where we want to let the system resolve e.g. `node`, but also want to have a nice Volta error to show if that fails (instead of the generic "command not found" error that the OS would otherwise provide. I'll update the name, I like `on_failure` as it's more descriptive of what the field is for.

I like the idea of adding a doc comment in spots where it isn't trivially obvious what something is for, so I'll add one here.